### PR TITLE
chore: setup devcontainer for CLI contributions

### LINKED ISSUE #6904: Easier start for Radius CLI contributors
### Area for Improvement (required)

Developing contributions for the Radius CLI

### Observed behavior (required)

When making contributions or working on bugs the developer needs to setup the necessary environment manually. 
Debugging the CLI from scratch might also be a barrier for new contributors

### Desired behavior (required)

It would be an improvment if the following parts exists:

- devcontainer configuration containing the basic setup for working with the CLI
- a debug launch configuration that allows the contributor to directly exectute a command of the CLI

### Proposed Fix (optional)

- Add a new devcontainer configuration
- Enhance the `launch.json` with a debug configuration targeting the CLI

### rad Version (required)

n/a as working with local build when contributing

### Operating system (required)

Linux (devcontainer)

## Additional context

n/a

[AB#10643](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/10643)

### PR DESCRIPTION
**Pull request by lechnerc77**
# Description

- This PR adds a new devconatiner configuration with the basic setup to contribute to the radius project esp. the CLI. It contains the necessary tools to allow a quick starting point for contributions and bug analysis and fixing. This complements the setup already available for the tutorials
- In order to make the debug experience easier, a launch configuration was added to enable a direct entry into the CLI commands

## Type of change

- This pull request is a minor refactor, code cleanup, test improvement, or other maintenance task and doesn't change the functionality of Radius.

Fixes: #6904

### Summary
🐳🛠️🐛

This pull request adds new VS Code and devcontainer settings to improve the development and debugging experience for the Radius CLI project. It allows contributors to run and debug the `radius` command with different arguments, and to use various tools and features in a consistent and isolated environment.

> _We are the coders of the Radius CLI_
> _We build and debug with Go and VS Code_
> _We unleash the power of devcontainer_
> _We run `init --full` to start the adventure_

### Walkthrough
*  Add a devcontainer configuration for contributors ([link](https://github.com/radius-project/radius/pull/6905/files?diff=unified&w=0#diff-935f2b2e55e0ab1017fd1c544b9191e499e90e473ece926f8e65e95ddfbb1b1eL1-R30))
* Add a launch configuration for debugging the Radius CLI ([link](https://github.com/radius-project/radius/pull/6905/files?diff=unified&w=0#diff-bd5430ee7c51dc892a67b3f2829d1f5b6d223f0fd48b82322cfd45baf9f5e945R8-R16), [link](https://github.com/radius-project/radius/pull/6905/files?diff=unified&w=0#diff-bd5430ee7c51dc892a67b3f2829d1f5b6d223f0fd48b82322cfd45baf9f5e945L92-R107))

### CODE REVIEW

File: .devcontainer/contributor/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
+{
+    "name": "Radius CLI - Contributor",
```
Comment (rynowak) [inline .devcontainer/contributor/devcontainer.json:2]: Something I wonder - does this need to be specific to the CLI? It feels like this is useful for any kind of contribution.

File: .devcontainer/contributor/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    "image": "mcr.microsoft.com/devcontainers/base:ubuntu",
+    "features": {
+        "ghcr.io/devcontainers/features/go:1": {
+            "version": "1.21"
```
Comment (rynowak) [inline .devcontainer/contributor/devcontainer.json:6]: Would it be OK for us to un-pin this? Go's tooling has been rock-solid for us across a variety of versions.

We already update `go.mod` when we need to. I'd rather not have a second thing to update.

File: .devcontainer/contributor/devcontainer.json [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        },
+        "ghcr.io/guiyomh/features/golangci-lint:0": {},
```
Comment (rynowak) [inline .devcontainer/contributor/devcontainer.json:8]: It looks like the Go feature also includes `golangci-lint`.

File: .devcontainer/contributor/devcontainer.json [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+            "moby": true
+        }
+    },
+    "customizations": {
+        "vscode": {
+            "extensions": [
+                "redhat.vscode-yaml",
+                "streetsidesoftware.code-spell-checker",
+                "golang.go",
+                "ms-azuretools.vscode-bicep",
```
Comment (rynowak) [inline .devcontainer/contributor/devcontainer.json:25]:
```suggestion
                "ms-azuretools.rad-vscode-bicep",
```

Until we converge on the official Bicep compiler, this needs to be our fork.

File: .vscode/launch.json [hunk 5]
```diff
@@ -4,6 +4,15 @@
...
     "configurations": [
+        {
+            "name": "Debug radius cli",
+            "type": "go",
+            "request": "launch",
+            "mode": "debug",
+            "program": "${workspaceFolder}/cmd/rad",
+            "args": "${input:cliArgs}",
+            "console": "integratedTerminal",
+        },
```
Comment (rynowak) [inline .vscode/launch.json:15]: Did you see the target defined here: https://github.com/radius-project/radius/pull/6905/files#diff-bd5430ee7c51dc892a67b3f2829d1f5b6d223f0fd48b82322cfd45baf9f5e945R64?

Is there a reason why we need two targets that both launch the CLI?

File: .devcontainer/contributor/devcontainer.json [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        "ghcr.io/devcontainers/features/docker-in-docker:2": {
+            "version": "latest",
+            "moby": true
+        }
+    },
+    "customizations": {
+        "vscode": {
+            "extensions": [
+                "redhat.vscode-yaml",
+                "streetsidesoftware.code-spell-checker",
```
Comment (rynowak) [inline .devcontainer/contributor/devcontainer.json:23]: What options do users have to install their own extensions when using a codespace/devcontainer? I'm asking because this isn't something we've adopted/documented as part of the project. Likewise `eamodio.gitlens`.

**Comment by rynowak** [comment]
Thanks for getting this started @lechnerc77 - I think this is generally on the right track and a good thing to add. Left some questions for you.

File: .devcontainer/contributor/devcontainer.json (hunk 1 above)
Comment (lechnerc77) [reply .devcontainer/contributor/devcontainer.json:2]: You are right. I will rephrase it to Radius - Contributor

File: .devcontainer/contributor/devcontainer.json (hunk 6 above)
Comment (lechnerc77) [reply .devcontainer/contributor/devcontainer.json:23]: I can add that one. In general, after the devcontainer/codespace started you can reinstall the additional extensions. However, as soon as you do a rebuild they are gone, so you must reinstall it

File: .devcontainer/contributor/devcontainer.json (hunk 3 above)
Comment (lechnerc77) [reply .devcontainer/contributor/devcontainer.json:8]: Cross checked, you are right -> removing the additional feature

File: .devcontainer/contributor/devcontainer.json (hunk 6 above)
Comment (lechnerc77) [reply .devcontainer/contributor/devcontainer.json:23]: They can install their own extension inside the container. However after a rebuild they are gone. 
I reduced the list to the minimum. Let me know if I should add/remove other extensions

File: .vscode/launch.json (hunk 5 above)
Comment (lechnerc77) [reply .vscode/launch.json:15]: Yes I saw the generic launch command. My motivation was the following scenario: 
I am a new contributor and I pick up an iccues that states an error e.g. in the `rad version` command (not super realistic, but as an example). Now I want to debug the command. I found the `version.go` file but now how to easily debug the command without any "detours". 
The added launch command should hopefull yhelp here as it clearly targets the debug scenario. The flow would be:

1. Set a breakpoint where you think it is needed
2. Start the "Debug Radius CLI" configuration
Screenshot 2023-12-05 081019

3.  Then you get prompted by VSCode for the command you want to execute (defaulted, to get an idea what to enter and with some additional tooltip) and you enter the command you want to execute
Screenshot 2023-12-05 081059

4. You can start debugging

Does that flow make sense to you? 

I think the other target remains useful as it is more generic, but maybe too generic for contrinbutors starting their journey. Waht do you think?

**Comment by lechnerc77** [comment]
> Thanks for getting this started @lechnerc77 - I think this is generally on the right track and a good thing to add. Left some questions for you.

@rynowak Thanks for coming back that quick and the helpful review comments. I changed the configuration in the places you highlighted. 
I also added some more details on the additional debug configuration.

File: .devcontainer/contributor/devcontainer.json [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        "ghcr.io/devcontainers/features/go:1": {},
+        "ghcr.io/devcontainers/features/git:1": {},
+        "ghcr.io/devcontainers/features/github-cli:1": {},
+        "ghcr.io/devcontainers/features/kubectl-helm-minikube:1": {
+            "minikube": "none"
+        },
+        "ghcr.io/devcontainers/features/docker-in-docker:2": {
+            "version": "latest",
+            "moby": true
+        }
```
Comment (ytimocin) [inline .devcontainer/contributor/devcontainer.json:14]: Just for my own understanding, what do the numbers represent here?

Comment (lechnerc77) [reply .devcontainer/contributor/devcontainer.json:14]: The numbers are the (major) versions of the *features* as defined here <https://containers.dev/implementors/features/#devcontainer-json-properties>.

The version of the content of the features is usually specified via the options as value to the `version` key.

File: .vscode/launch.json (hunk 5 above)
Comment (rynowak) [reply .vscode/launch.json:15]: > I think the other target remains useful as it is more generic, but maybe too generic for contrinbutors starting their journey. Waht do you think?

They are both debugging commands. The new one you added prompts for args, so I think it's also useful 👍 

Could you move them to be next to each other in the file name name them similar things. For example:

- Debug rad CLI 
- Debug rad CLI (prompt for args)

Comment (rynowak) [reply .vscode/launch.json:15]: Can you also add `"preLaunchTask": "Build Radius (all)"` to this? That will help with failure reporting.

The experience for understanding a build error is pretty bad without this.

**Comment by rynowak** [comment]
Thanks @lechnerc77 - this is looking really good. I had some very minor feedback about the additional debugging target. Could you take a look? 

After that I think this is ready to go in.

**Review by rynowak — APPROVED** [review]
Awesome! 😍 thank you so much. Approving test runs.

**Comment by rynowak** [comment]
Rerunning the failed job, there's no way it's something you caused with your PR.

**Comment by rynowak** [comment]
@lechnerc77 thanks!
